### PR TITLE
Corrected 'List' type struct definitions

### LINKED ISSUE #2220: All machine readable output "Lists" should use Kind: List
[kind/Enhancement]

## Which functionality do you think we should update/improve?

There are some json machine output commands that are not using the correct List for example:

- `odo catalog list services` uses ServiceTypeList
- `odo catalog list components` uses ComponentTypeList
- `odo application list` uses ApplicationList

These should be `List`

## Why is this needed? 

To follow Kubernetes standards.

### PR DESCRIPTION
**Pull request by dev-gaur**
**What type of PR is this?**
/kind cleanup

**Which issue(s) this PR fixes**:
Fixes #2220

### CODE REVIEW

File: pkg/catalog/types.go [hunk 1]
```diff
@@ -7,9 +7,9 @@ import (
 
 // ComponentType is the main struct for catalog components
 type ComponentType struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-	Spec              ComponentSpec `json:"spec,omitempty"`
+	metav1.TypeMeta		`json:",inline"`
```
Comment (dharmit) [inline pkg/catalog/types.go:10]: Formatting is not proper here. It's an issue in other files as well. Travis checks are failing for the same reason:
```sh
$ make validate
./scripts/check-gofmt.sh
gofmt ERROR - These files are not formated by gofmt:
./pkg/odo/cli/catalog/util/util.go
./pkg/catalog/types.go
./pkg/machineoutput/services.go
./pkg/machineoutput/types.go
./pkg/service/types.go
Makefile:65: recipe for target 'gofmt' failed
make: *** [gofmt] Error 1
The command "make validate" exited with 2.

```
You can fix this with the go fmt tool.

Comment (dev-gaur) [reply pkg/catalog/types.go]: Thanks for pointing that out. Changes made.

**Comment by dharmit** [comment]
[4.3 failure](https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3242/pull-ci-openshift-odo-master-v4.3-integration-e2e/648) looks like a flake (#3243):
```
[odo] I0526 09:56:35.469743   27355 utils.go:82] Waiting for rollout to finish: 0 of 1 updated replicas are available...
[odo]  ✗  Applying configuration [5m]
[odo]  ✗  Failed To Update Config To Component Deployed.
[odo] Error: unable to update DeploymentConfig for hkeck component: unable to update the current DeploymentConfig hkeck-ydtxa: unable to wait for DeploymentConfig hkeck-ydtxa to update: Timed out waiting for annotation to update
Deleting project: ztwcxauxvu
Running odo with args [odo project delete ztwcxauxvu -f]
[odo] I0526 10:01:26.010170   28600 application.go:49] Unable to list Service Catalog instances: unable to list ServiceInstances: serviceinstances.servicecatalog.k8s.io is forbidden: User "developer" cannot list resource "serviceinstances" in API group "servicecatalog.k8s.io" in the namespace "ztwcxauxvu"
[odo] This project contains the following applications, which will be deleted
[odo] Application ydtxa
[odo] I0526 10:01:26.077611   28600 occlient.go:2821] Getting DeploymentConfig: hkeck-ydtxa
[odo] I0526 10:01:26.108298   28600 component.go:987] Source for component hkeck is  (local)
[odo] I0526 10:01:26.108330   28600 url.go:331] Listing routes with label selector: app.kubernetes.io/part-of=ydtxa,app.kubernetes.io/instance=hkeck
[odo] I0526 10:01:26.108339   28600 occlient.go:2680] Listing routes with label selector: app.kubernetes.io/part-of=ydtxa,app.kubernetes.io/instance=hkeck
[odo] I0526 10:01:26.133047   28600 preference.go:188] The path for preference file is /tmp/988743992/config.yaml
[odo] I0526 10:01:26.171366   28600 occlient.go:2821] Getting DeploymentConfig: hkeck-ydtxa
[odo] This application has following components that will be deleted
[odo] I0526 10:01:26.238725   28600 occlient.go:2821] Getting DeploymentConfig: hkeck-ydtxa
[odo] I0526 10:01:26.257334   28600 component.go:987] Source for component hkeck is  (local)
[odo] I0526 10:01:26.257359   28600 url.go:331] Listing routes with label selector: app.kubernetes.io/part-of=ydtxa,app.kubernetes.io/instance=hkeck
[odo] I0526 10:01:26.257367   28600 occlient.go:2680] Listing routes with label selector: app.kubernetes.io/part-of=ydtxa,app.kubernetes.io/instance=hkeck
[odo] I0526 10:01:26.277930   28600 preference.go:188] The path for preference file is /tmp/988743992/config.yaml
[odo] I0526 10:01:26.318392   28600 occlient.go:2821] Getting DeploymentConfig: hkeck-ydtxa
[odo] component named hkeck
[odo] I0526 10:01:26.383512   28600 url.go:331] Listing routes with label selector: app.kubernetes.io/part-of=ydtxa,app.kubernetes.io/instance=hkeck
[odo] I0526 10:01:26.383522   28600 occlient.go:2680] Listing routes with label selector: app.kubernetes.io/part-of=ydtxa,app.kubernetes.io/instance=hkeck
[odo] I0526 10:01:26.407603   28600 preference.go:188] The path for preference file is /tmp/988743992/config.yaml
[odo] This component has following urls that will be deleted with component
[odo] URL named example-1 with host example-1-ydtxa-ztwcxauxvu.apps.ci-op-xytfli4p-25553.origin-ci-int-aws.dev.rhcloud.com having protocol http at port 8080
[odo] No services / could not get services
[odo] I0526 10:01:26.423478   28600 project.go:136] unable to list services: unable to list ServiceInstances: serviceinstances.servicecatalog.k8s.io is forbidden: User "developer" cannot list resource "serviceinstances" in API group "servicecatalog.k8s.io" in the namespace "ztwcxauxvu"
[odo]  ⚠  Warning! Projects are deleted from the cluster asynchronously. Odo does its best to delete the project. Due to multi-tenant clusters, the project may still exist on a different node.
[odo] I0526 10:01:26.450670   28600 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
[odo]  ✓  Deleted project : ztwcxauxvu
Deleting dir: /tmp/988743992
• Failure [334.550 seconds]
odo component command tests
/go/src/github.com/openshift/odo/tests/integration/cmd_cmp_test.go:13
  odo component delete should clean owned resources
  /go/src/github.com/openshift/odo/tests/integration/component.go:847
    should delete the component and the owned resources with wait flag [It]
    /go/src/github.com/openshift/odo/tests/integration/component.go:882
    No future change is possible.  Bailing out early after 300.492s.
      
    Running odo with args [odo push --context /tmp/988743992]
    Expected
        <int>: 1
    to match exit code:
        <int>: 0
    /go/src/github.com/openshift/odo/tests/helper/helper_run.go:34 
```
[4.4 failure](https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3242/pull-ci-openshift-odo-master-v4.4-integration-e2e/633) looks like infra issue:
```
 error: some steps failed:
  * could not run steps: step integration-e2e failed: template pod "integration-e2e" failed: the pod ci-op-vz99wfqp/integration-e2e failed after 2m42s (failed containers: setup): ContainerFailed one or more containers exited
Container setup exited with code 1, reason Error
---
Installing from release registry.svc.ci.openshift.org/ci-op-vz99wfqp/release@sha256:46e5c90db210a7766c1d6a5355641b0b6c5bfc0f77b9869dc00b74ac96a0b724
AWS region: us-east-2 (zones: us-east-2a us-east-2b)
level=info msg="Credentials loaded from the \"default\" profile in file \"/etc/openshift-installer/.awscred\""
level=info msg="Consuming Install Config from target directory"
level=warning msg="Found override for release image. Please be warned, this is not advised"
level=info msg="Consuming Worker Machines from target directory"
level=info msg="Consuming Openshift Manifests from target directory"
level=info msg="Consuming Master Machines from target directory"
level=info msg="Consuming Common Manifests from target directory"
level=info msg="Consuming OpenShift Install (Manifests) from target directory"
level=info msg="Credentials loaded from the \"default\" profile in file \"/etc/openshift-installer/.awscred\""
level=fatal msg="failed to fetch Cluster: failed to fetch dependency of \"Cluster\": failed to generate asset \"Platform Permissions Check\": validate AWS credentials: checking install permissions: error simulating policy: Throttling: Rate exceeded\n\tstatus code: 400, request id: 1af24733-91a4-4d2d-ac30-e568d3560c13"
--- 
```

**Comment by dharmit** [comment]
My understanding of the issue this PR is trying to fix is that we intend to replace things like `ServiceTypeList`, `ComponentTypeList`, etc. with a type called `List`. Correct me if that's wrong.

I have two questions related to the code and they might be naive:
1. Should we not create a new type called `List` for this? Or would that be over-engineering?
2. Why are we replacing `ObjectMeta` with `ListMeta`?

@cdrage @girishramnani

**Comment by dev-gaur** [comment]
> My understanding of the issue this PR is trying to fix is that we intend to replace things like `ServiceTypeList`, `ComponentTypeList`, etc. with a type called `List`. Correct me if that's wrong.
> 
> I have two questions related to the code and they might be naive:
> 
> 1. Should we not create a new type called `List` for this? Or would that be over-engineering?
IMHO, yes.
> 2. Why are we replacing `ObjectMeta` with `ListMeta`?
Because it was supposed to be ListMeta only for List types. It was wrong to put ObjectMeta there. You can  refer to any other kubernetes/openshift related repository. All the List types use ListMeta instead of OjbectMeta.

File: pkg/machineoutput/types.go [hunk 2]
```diff
@@ -55,6 +53,6 @@ func OutputError(machineOutput interface{}) {
 }
 
 // MarshalJSONIndented returns indented json representation of obj
-func MarshalJSONIndented(obj interface{}) ([]byte, error) {
-	return json.MarshalIndent(obj, "", "    ")
+func marshalJSONIndented(obj interface{}) ([]byte, error) {
```
Comment (mik-dass) [inline pkg/machineoutput/types.go:56]: `marshalJSONIndented` in the comment too

**Comment by mik-dass** [comment]
You need to replace the strings in the integration tests too 
https://travis-ci.com/github/openshift/odo/jobs/339935962#L574

**Comment by girishramnani** [comment]
@dgolovin please confirm that is changes are ok as its changing the machine readable output.

**Comment by dgolovin** [comment]
@girishramnani that should be fine, because vscode extension does not check type of the list, it just using `items` properties assuming any kind of list has it.

**Comment by dharmit** [comment]
travis failure might be a flake:
```
Created dir: /tmp/290352664
Creating a new project: ouhqnkplip
Running odo with args [odo project create ouhqnkplip -w -v4]
[odo]  •  Waiting for project to come up  ...
[odo] I0611 12:45:12.536848   24306 occlient.go:543] Status of creation of project ouhqnkplip is Active
[odo] I0611 12:45:12.536888   24306 occlient.go:548] Project ouhqnkplip now exists
[odo] I0611 12:45:12.539289   24306 occlient.go:583] Status of creation of service account &ServiceAccount{ObjectMeta:{default  ouhqnkplip /api/v1/namespaces/ouhqnkplip/serviceaccounts/default 641a4e58-abe1-11ea-8e87-42010a1e0043 3509 0 2020-06-11 12:45:12 +0000 UTC <nil> <nil> map[] map[openshift.io/create-dockercfg-secrets.pending-token:default-token-rt8qn] [] []  []},Secrets:[]ObjectReference{},ImagePullSecrets:[]LocalObjectReference{},AutomountServiceAccountToken:nil,} is ready
 ✓  Waiting for project to come up [413ms]
[odo]  ✓  Project 'ouhqnkplip' is ready for use
[odo]  ✓  New project created and now using project: ouhqnkplip
[odo] I0611 12:45:12.542512   24306 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
Running odo with args [odo component create nodejs nodejs --app nodeapp --project ouhqnkplip --context /tmp/290352664]
[odo] I0611 12:45:12.635174   24315 preference.go:188] The path for preference file is /tmp/290352664/config.yaml
[odo] I0611 12:45:12.656959   24315 preference.go:188] The path for preference file is /tmp/290352664/config.yaml
[odo]  ✗  invalid configuration: no configuration has been provided
Deleting project: ouhqnkplip
Running odo with args [odo project delete ouhqnkplip -f]
[odo] I0611 12:45:12.958860   24323 application.go:49] Unable to list Service Catalog instances: unable to list ServiceInstances: serviceinstances.servicecatalog.k8s.io is forbidden: User "developer" cannot list serviceinstances.servicecatalog.k8s.io in the namespace "ouhqnkplip": no RBAC policy matched
[odo]  ✓  Deleted project : ouhqnkplip
[odo]  ⚠  Warning! Projects are deleted from the cluster asynchronously. Odo does its best to delete the project. Due to multi-tenant clusters, the project may still exist on a different node.
[odo] I0611 12:45:12.981186   24323 odo.go:72] Could not get the latest release information in time. Never mind, exiting gracefully :)
Deleting dir: /tmp/290352664
• Failure [1.086 seconds]
odo storage command tests
/home/travis/gopath/src/github.com/openshift/odo/tests/integration/cmd_storage_test.go:13
  when using storage command with default flag values
  /home/travis/gopath/src/github.com/openshift/odo/tests/integration/cmd_storage_test.go:53
    should add a storage, list and delete it [It]
    /home/travis/gopath/src/github.com/openshift/odo/tests/integration/cmd_storage_test.go:54
    No future change is possible.  Bailing out early after 0.132s.
    Running odo with args [odo component create nodejs nodejs --app nodeapp --project ouhqnkplip --context /tmp/290352664]
    Expected
        <int>: 1
    to match exit code:
        <int>: 0
    /home/travis/gopath/src/github.com/openshift/odo/tests/helper/helper_run.go:34

```

**Comment by mik-dass** [comment]
@dev-gaur rebase needed

**Comment by maysunfaisal** [comment]
/hold
until PR 3291 is merged

**Pull request merged**